core/internal/runbranch/rewind.go

```diff
@@ -0,0 +1,168 @@
+package runbranch
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/simplejsonext"
+	"github.com/wandb/wandb/core/internal/filestream"
+	"github.com/wandb/wandb/core/internal/gql"
+	"github.com/wandb/wandb/core/pkg/service"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type RewindBranch struct {
+	ctx    context.Context
+	client graphql.Client
+	branch BranchPoint
+}
+
+func NewRewindBranch(
+	ctx context.Context,
+	client graphql.Client,
+	runid string,
+	metricName string,
+	metricValue float64,
+) *RewindBranch {
+	return &RewindBranch{
+		ctx:    ctx,
+		client: client,
+		branch: BranchPoint{
+			RunID:       runid,
+			MetricName:  metricName,
+			MetricValue: metricValue,
+		},
+	}
+}
+
+//gocyclo:ignore
+func (rb RewindBranch) GetUpdates(
+	params *RunParams,
+	runpath RunPath,
+) (*RunParams, error) {
+	if rb.branch.RunID != runpath.RunID {
+		err := fmt.Errorf("rewind run id %s does not match run id %s", rb.branch.RunID, runpath.RunID)
+		info := &service.ErrorInfo{
+			Code:    service.ErrorInfo_USAGE,
+			Message: err.Error(),
+		}
+		return nil, &BranchError{Err: err, Response: info}
+	}
+
+	if rb.branch.MetricName != "_step" {
+		err := fmt.Errorf("rewind only supports `_step` metric name currently")
+		info := &service.ErrorInfo{
+			Code:    service.ErrorInfo_UNSUPPORTED,
+			Message: err.Error(),
+		}
+		return nil, &BranchError{Err: err, Response: info}
+	}
+
+	response, err := gql.RewindRun(
+		rb.ctx,
+		rb.client,
+		runpath.RunID,
+		utils.NilIfZero(runpath.Entity),
+		utils.NilIfZero(runpath.Project),
+		rb.branch.MetricName,
+		rb.branch.MetricValue,
+	)
+	if err != nil {
+		info := &service.ErrorInfo{
+			Code:    service.ErrorInfo_COMMUNICATION,
+			Message: fmt.Sprintf("failed to rewind run: %s", err),
+		}
+		return nil, &BranchError{Err: err, Response: info}
+	}
+
+	if response.GetRewindRun() == nil || response.GetRewindRun().GetRewoundRun() == nil {
+		info := &service.ErrorInfo{
+			Code:    service.ErrorInfo_COMMUNICATION,
+			Message: "failed to rewind run: run not found",
+		}
+		return nil, &BranchError{Err: fmt.Errorf("run not found"), Response: info}
+	}
+
+	r := &RunParams{}
+	r.Merge(params)
+
+	data := response.GetRewindRun().GetRewoundRun()
+
+	if data.GetId() != "" {
+		r.StorageID = data.GetId()
+	}
+
+	if data.GetName() != "" {
+		r.RunID = data.GetName()
+	}
```

Comment:
these wouldn't need to be guarded by `if` statements if this function wasn't also merging values

Reply:
ah it's because i though maybe we should fail on empty values, but now i think we just merge it